### PR TITLE
[#432] Add Groq Provider - chat completions

### PR DESCRIPTION
**Pull request by aidando73**
# What does this PR do?

Contributes towards issue (#432)

- Groq text chat completions
- Streaming
- All the sampling params that Groq supports

A lot of inspiration taken from @mattf's good work at https://github.com/meta-llama/llama-stack/pull/355

**What this PR does not do**

- Tool calls (Future PR)
- Adding llama-guard model
- See if we can add embeddings

### PR Train

- https://github.com/meta-llama/llama-stack/pull/609 👈 
- https://github.com/meta-llama/llama-stack/pull/630

## Test Plan

<details>

<summary>Environment</summary>

```bash
export GROQ_API_KEY=<api_key>

wget https://raw.githubusercontent.com/aidando73/llama-stack/240e6e2a9c20450ffdcfbabd800a6c0291f19288/build.yaml
wget https://raw.githubusercontent.com/aidando73/llama-stack/92c9b5297f9eda6a6e901e1adbd894e169dbb278/run.yaml

# Build and run environment
pip install -e . \
&& llama stack build --config ./build.yaml --image-type conda \
&& llama stack run ./run.yaml \
  --port 5001
```

</details>

<details>

<summary>Manual tests</summary>

Using this jupyter notebook to test manually: https://github.com/aidando73/llama-stack/blob/2140976d76ee7ef46025c862b26ee87585381d2a/hello.ipynb

Use this code to test passing in the api key from provider_data

```
from llama_stack_client import LlamaStackClient

client = LlamaStackClient(
    base_url="http://localhost:5001",
)

response = client.inference.chat_completion(
    model_id="Llama3.2-3B-Instruct",
    messages=[
        {"role": "user", "content": "Hello, world client!"},
    ],
    # Test passing in groq_api_key from the client
    # Need to comment out the groq_api_key in the run.yaml file
    x_llama_stack_provider_data='{"groq_api_key": "<api-key>"}',
    # stream=True,
)
response
```

</details>

<details>
<summary>Integration</summary>

`pytest llama_stack/providers/tests/inference/test_text_inference.py -v -k groq`

(run in same environment)

```
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_model_list[llama_3b-groq] PASSED                 [  6%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion[llama_3b-groq] SKIPPED (Other inf...) [ 12%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion_structured_output[llama_3b-groq] SKIPPED [ 18%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_non_streaming[llama_3b-groq] PASSED [ 25%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_structured_output[llama_3b-groq] SKIPPED (Ot...) [ 31%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_streaming[llama_3b-groq] PASSED  [ 37%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling[llama_3b-groq] SKIPPED [ 43%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling_streaming[llama_3b-groq] SKIPPED [ 50%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_model_list[llama_8b-groq] PASSED                 [ 56%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion[llama_8b-groq] SKIPPED (Other inf...) [ 62%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion_structured_output[llama_8b-groq] SKIPPED [ 68%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_non_streaming[llama_8b-groq] PASSED [ 75%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_structured_output[llama_8b-groq] SKIPPED (Ot...) [ 81%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_streaming[llama_8b-groq] PASSED  [ 87%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling[llama_8b-groq] SKIPPED [ 93%]
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling_streaming[llama_8b-groq] SKIPPED [100%]

======================================= 6 passed, 10 skipped, 160 deselected, 7 warnings in 2.05s ========================================
```
</details>

<details>
<summary>Unit tests</summary>

`pytest llama_stack/providers/tests/inference/groq/ -v`

```
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_sets_model PASSED            [  5%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_converts_user_message PASSED [ 10%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_converts_system_message PASSED [ 15%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_converts_completion_message PASSED [ 20%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_does_not_include_logprobs PASSED [ 25%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_does_not_include_response_format PASSED [ 30%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_does_not_include_repetition_penalty PASSED [ 35%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_includes_stream PASSED       [ 40%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_n_is_1 PASSED                [ 45%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_if_max_tokens_is_0_then_it_is_not_included PASSED [ 50%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_includes_max_tokens_if_set PASSED [ 55%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_includes_temperature PASSED  [ 60%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertChatCompletionRequest::test_includes_top_p PASSED        [ 65%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertNonStreamChatCompletionResponse::test_returns_response PASSED [ 70%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertNonStreamChatCompletionResponse::test_maps_stop_to_end_of_message PASSED [ 75%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertNonStreamChatCompletionResponse::test_maps_length_to_end_of_message PASSED [ 80%]
llama_stack/providers/tests/inference/groq/test_groq_utils.py::TestConvertStreamChatCompletionResponse::test_returns_stream PASSED [ 85%]
llama_stack/providers/tests/inference/groq/test_init.py::TestGroqInit::test_raises_runtime_error_if_config_is_not_groq_config PASSED [ 90%]
llama_stack/providers/tests/inference/groq/test_init.py::TestGroqInit::test_returns_groq_adapter PASSED                            [ 95%]
llama_stack/providers/tests/inference/groq/test_init.py::TestGroqConfig::test_api_key_defaults_to_env_var PASSED                   [100%]

==================================================== 20 passed, 11 warnings in 0.08s =====================================================
```

</details>

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Ran pre-commit to handle lint / formatting issues.
- [x] Read the [contributor guideline](https://github.com/meta-llama/llama-stack/blob/main/CONTRIBUTING.md),
      Pull Request section?
- [x] Updated relevant documentation
- [x] Wrote necessary unit or integration tests.

### CODE REVIEW

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    stop -> model hit a natural stop point or a provided stop sequence
+    length -> maximum number of tokens specified in the request was reached
+    tool_calls -> model called a tool
+    """
+    if finish_reason == "stop":
+        return StopReason.end_of_turn
+    elif finish_reason == "length":
+        return StopReason.end_of_message
+    elif finish_reason == "tool_calls":
+        raise NotImplementedError("tool_calls is not supported yet")
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:106]: Users won't be able to hit this error yet since they can't pass tools as a parameter

File: llama_stack/providers/tests/inference/test_text_inference.py [hunk 2]
```diff
@@ -295,6 +295,13 @@ async def test_chat_completion_with_tool_calling(
         sample_messages,
         sample_tool_definition,
     ):
+        inference_impl, _ = inference_stack
+        provider = inference_impl.routing_table.get_provider_impl(inference_model)
+        if provider.__provider_spec__.provider_type in (
+            "remote::groq",
+        ):
+            pytest.skip(provider.__provider_spec__.provider_type + " doesn't support tool calling yet")
+
```
Comment (aidando73) [inline llama_stack/providers/tests/inference/test_text_inference.py:304]: I’m planning to handle tool calls in a separate PR since there are edge cases I want to cover properly.

But lmk if you want me to include it within this PR

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    if request.sampling_params.repetition_penalty:
+        # groq supports frequency_penalty, but frequency_penalty and sampling_params.repetition_penalty
+        # seem to have different semantics
+        # frequency_penalty defaults to 0 is a float between -2.0 and 2.0
+        # repetition_penalty defaults to 1 and is often set somewhere between 1.0 and 2.0
+        # so we exclude it for now
+        warnings.warn("repetition_penalty is not supported")
+    
+    if request.tools:
+        warnings.warn("tools are not supported yet")
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:52]: I’m planning to handle tool calls in a separate PR since there are edge cases I want to cover properly.

But lmk if you want me to include it within this PR

File: llama_stack/providers/remote/inference/groq/config.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+from typing import Optional
+
+from llama_models.schema_utils import json_schema_type
+from pydantic import BaseModel, Field
+
+
+@json_schema_type
+class GroqConfig(BaseModel):
+    api_key: Optional[str] = Field(
+        default=None,
```
Comment (mattf) [inline llama_stack/providers/remote/inference/groq/config.py:16]: nit - the groq library will read GROQ_API_KEY env (https://github.com/groq/groq-python/blob/main/src/groq/_client.py#L86), consider adding a comment here so people in the LS codebase know this expectation

File: llama_stack/providers/remote/inference/groq/groq.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            ToolPromptFormat
+        ] = None,  # API default is ToolPromptFormat.json, we default to None to detect user input
+        stream: Optional[bool] = False,
+        logprobs: Optional[LogProbConfig] = None,
+    ) -> Union[
+        ChatCompletionResponse, AsyncIterator[ChatCompletionResponseStreamChunk]
+    ]:
+
+        if model_id == "llama-3.2-3b-preview":
+            warnings.warn(
```
Comment (mattf) [inline llama_stack/providers/remote/inference/groq/groq.py:109]: very user friendly +1

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    return CompletionCreateParams(
+        model=request.model,
+        messages=[_convert_message(message) for message in request.messages],
+        logprobs=None,
+        frequency_penalty=None,
+        stream=request.stream,
+        # Groq only supports n=1 at the time of writing
```
Comment (mattf) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:72]: the LS structures only support responses w/ 1 choice, so the fact that groq only supports n=1 is moot. you should even skip passing the param around.

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        return StopReason.end_of_turn
+    elif finish_reason == "length":
+        return StopReason.end_of_message
+    elif finish_reason == "tool_calls":
+        raise NotImplementedError("tool_calls is not supported yet")
+    else:
+        raise ValueError(f"Invalid finish reason: {finish_reason}")
+
+
+async def convert_chat_completion_response_stream(
```
Comment (mattf) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:127]: in some other PR, we should merge this into the general util module.

**Review by mattf — COMMENTED** [review]
looking good!

File: llama_stack/providers/remote/inference/groq/config.py (hunk 4 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/config.py:17]: Done

File: llama_stack/providers/remote/inference/groq/groq_utils.py (hunk 7 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/groq_utils.py:125]: Do you think this function is too coupled to Groq types to be used as general util functions? E.g., this one takes in a `ChatCompletion` from `groq.types.chat.chat_completion`

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    request: ChatCompletionRequest,
+) -> CompletionCreateParams:
+    """
+    Convert a ChatCompletionRequest to a Groq API-compatible dictionary.
+    Warns client if request contains unsupported features.
+    """
+
+    if request.logprobs:
+        # Groq doesn't support logprobs at the time of writing
+        warnings.warn("logprobs are not supported yet")
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:49]: image

https://console.groq.com/docs/api-reference#chat

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+    if request.response_format:
+        # Groq's JSON mode is beta at the time of writing
+        warnings.warn("response_format is not supported yet")
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:53]: JSON mode still in beta for Groq:

image

https://console.groq.com/docs/text-chat#json-mode-object-object

File: llama_stack/providers/remote/inference/groq/groq.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        tool_choice: Optional[ToolChoice] = ToolChoice.auto,
+        tool_prompt_format: Optional[
+            ToolPromptFormat
+        ] = None,  # API default is ToolPromptFormat.json, we default to None to detect user input
+        stream: Optional[bool] = False,
+        logprobs: Optional[LogProbConfig] = None,
+    ) -> Union[
+        ChatCompletionResponse, AsyncIterator[ChatCompletionResponseStreamChunk]
+    ]:
+
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq.py:107]:
```suggestion
model_id = self.get_provider_model_id(model_id)
```

File: llama_stack/providers/remote/inference/groq/groq.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        if model_id == "llama-3.2-3b-preview":
+            warnings.warn(
+                "Groq only contains a preview version for llama-3.2-3b-instruct. "
+                "Preview models aren't recommended for production use. "
+                "They can be discontinued on short notice."
+            )
+
+        model_id = self.get_provider_model_id(model_id)
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq.py:115]:
```suggestion
```

File: llama_stack/providers/remote/inference/groq/groq.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        tool_choice: Optional[ToolChoice] = ToolChoice.auto,
+        tool_prompt_format: Optional[
+            ToolPromptFormat
+        ] = None,  # API default is ToolPromptFormat.json, we default to None to detect user input
+        stream: Optional[bool] = False,
+        logprobs: Optional[LogProbConfig] = None,
+    ) -> Union[
+        ChatCompletionResponse, AsyncIterator[ChatCompletionResponseStreamChunk]
+    ]:
+model_id = self.get_provider_model_id(model_id)
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq.py:107]:
```suggestion
        model_id = self.get_provider_model_id(model_id)
```

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    if request.sampling_params.repetition_penalty:
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:55]:
```suggestion
    if request.sampling_params.repetition_penalty != 1.0:
```

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+    finish_reason: Literal["stop", "length", "tool_calls"]
+        - stop -> model hit a natural stop point or a provided stop sequence
+        - length -> maximum number of tokens specified in the request was reached
+        - tool_calls -> model called a tool
+    """
+    if finish_reason == "stop":
+        return StopReason.end_of_turn
+    elif finish_reason == "length":
+        return StopReason.end_of_message
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:118]: TODO - imo this should be out_of_tokens

Grok returns length if we hit the max_tokens specified in request or ran out of the model's maximum tokens.

File: llama_stack/providers/tests/inference/groq/test_groq_utils.py [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+
+        assert converted["stream"] is True
+
+    def test_n_is_1(self):
+        request = self._dummy_chat_completion_request()
+
+        converted = convert_chat_completion_request(request)
+
+        assert converted["n"] == 1
+
```
Comment (aidando73) [inline llama_stack/providers/tests/inference/groq/test_groq_utils.py:131]:
```suggestion
```

File: llama_stack/providers/remote/inference/groq/groq_utils.py (hunk 14 above)
Comment (aidando73) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:118]:
```suggestion
        return StopReason.out_of_tokens
```

File: llama_stack/providers/tests/inference/groq/test_groq_utils.py [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+        assert converted.completion_message.stop_reason == StopReason.end_of_turn
+
+    def test_maps_length_to_end_of_message(self):
+        response = self._dummy_chat_completion_response()
+        response.choices[0].finish_reason = "length"
+
+        converted = convert_chat_completion_response(response)
+
+        assert converted.completion_message.stop_reason == StopReason.end_of_message
```
Comment (aidando73) [inline llama_stack/providers/tests/inference/groq/test_groq_utils.py:189]:
```suggestion
    def test_maps_length_to_end_of_message(self):
        response = self._dummy_chat_completion_response()
        response.choices[0].finish_reason = "length"

        converted = convert_chat_completion_response(response)

        assert converted.completion_message.stop_reason == StopReason.out_of_tokens
```

File: llama_stack/providers/remote/inference/groq/config.py (hunk 4 above)
Comment (raghotham) [reply llama_stack/providers/remote/inference/groq/config.py:17]: Let's not rely on environment variables for code that we expect to run in llama-stack server. We would want to take in the api key as a config variable in run.yaml when we spin up the server

Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/config.py:17]: > We would want to take in the api key as a config variable in run.yaml when we spin up the server

@raghotham, I believe that's the behaviour at the moment. This is how fireworks and together define their configs:

https://github.com/meta-llama/llama-stack/blob/96e158eaac4aca62a62afeae40558e053627e547/llama_stack/providers/remote/inference/fireworks/config.py#L19-L22

https://github.com/meta-llama/llama-stack/blob/6765fd76fff516e654390ab9b21d74b6299ebd29/llama_stack/providers/remote/inference/together/config.py#L19-L22

And it's in the run.yaml that you can template in the environment variable:


wdyt?

File: llama_stack/providers/remote/inference/groq/__init__.py [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+# This source code is licensed under the terms described in the LICENSE file in
+# the root directory of this source tree.
+
+from llama_stack.apis.inference import Inference
+
+from .config import GroqConfig
+
+
+async def get_adapter_impl(config: GroqConfig, _deps) -> Inference:
+    # import dynamically so `llama stack build` does not fail due to missing dependencies
```
Comment (ashwinb) [inline llama_stack/providers/remote/inference/groq/__init__.py:13]: the correct comment should be `import dynamically so the import is used only when it is needed` :D

File: llama_stack/providers/remote/inference/groq/config.py (hunk 4 above)
Comment (ashwinb) [reply llama_stack/providers/remote/inference/groq/config.py:17]: yes @aidando73 this is correct. Note that both Together and Fireworks also support grabbing the api key from headers via the `NeedsProviderData` mixin. You can add that if you feel like it.

File: llama_stack/providers/remote/inference/groq/groq.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    build_model_alias_with_just_provider_model_id(
+        "llama-3.1-8b-instant",
+        CoreModelId.llama3_1_8b_instruct.value,
+    ),
+    build_model_alias(
+        "llama3-70b-8192",
+        CoreModelId.llama3_70b_instruct.value,
+    ),
+    build_model_alias(
+        "llama-3.3-70b-versatile",
```
Comment (ashwinb) [inline llama_stack/providers/remote/inference/groq/groq.py:58]: do you know what does this suffix indicate?

File: llama_stack/providers/remote/inference/groq/groq_utils.py [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        choice = chunk.choices[0]
+
+        # We assume there's only one finish_reason for the entire stream.
+        # We collect the last finish_reason
+        if choice.finish_reason:
+            stop_reason = _map_finish_reason_to_stop_reason(choice.finish_reason)
+
+        yield ChatCompletionResponseStreamChunk(
+            event=ChatCompletionResponseEvent(
+                event_type=next(event_types),
```
Comment (ashwinb) [inline llama_stack/providers/remote/inference/groq/groq_utils.py:149]: it seems unnecessary to refactor the event type generator separately. just maintain an index here or even more simply just repeat a small amount of code to send back a start chunk.

**Review by ashwinb — APPROVED** [review]
awesome, thank you very much!

**Comment by ashwinb** [comment]
(will merge after a couple small comments are addressed)

File: llama_stack/providers/remote/inference/groq/__init__.py (hunk 17 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/__init__.py]: Done

File: llama_stack/providers/remote/inference/groq/config.py (hunk 4 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/config.py:17]: Done - added the mixin

File: llama_stack/providers/remote/inference/groq/groq.py (hunk 18 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/groq.py:56]: I couldn't find anything online. @ricklamers @philass - could you provide any additional context here?

File: llama_stack/providers/remote/inference/groq/groq_utils.py (hunk 19 above)
Comment (aidando73) [reply llama_stack/providers/remote/inference/groq/groq_utils.py]: Done

**Comment by aidando73** [comment]
Ok @ashwinb I've addressed your comments

**Comment by cheesecake100201** [comment]
Can this PR be merged soon ?
@ashwinb @aidando73

**Comment by ashwinb** [comment]
Thanks for the reminder @cheesecake100201! Sorry was out on vacation and missed this one after coming back.

**Comment by aidando73** [comment]
@ashwinb I missed a variable name while refactoring could you PTAL https://github.com/meta-llama/llama-stack/pull/719 👈 ?
